File: circle/C++/circletest.cpp
#include<iostream>
#include<opencv2/opencv.hpp>
#include<vector>
using namespace cv;
using namespace std;

int main()
{
	Mat src, gray, dst;
	src = imread("images/01.jpg");
	if (src.empty())
	{
		printf("can not load image \n");
		return -1;
	}
	cvtColor(src, gray, CV_RGB2GRAY);
	medianBlur(gray, dst , 7);
	//imshow("src", src);
	//imshow("gray", gray);
	//imshow("dst", dst);


	vector<Vec3f> circles;
	HoughCircles(dst, circles, CV_HOUGH_GRADIENT, 1, 120, 100, 30, 30, 220);
	for (size_t i = 0; i < circles.size(); i++)
	{
		Vec3f c = circles[i];
		circle(src, Point(c[0], c[1]), c[2], Scalar(0, 0, 255), 3, CV_AA);
		circle(src, Point(c[0], c[1]), 2, Scalar(0, 0, 255), 2);
		cout << "a = " << c[0] << "  " << "b = " << c[1] << "  " << "r = " << c[2] << endl;
	}
	imshow("output", src);
	waitKey(99999);

	return 0;
}

File: circle/circle_test.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 29 20:35:21 2021

@author: thebin
"""

import cv2 as cv
import numpy as np

planets = cv.imread("images/01.jpg")
gay_img = cv.cvtColor(planets, cv.COLOR_BGRA2GRAY)
img = cv.medianBlur(gay_img, 7)  # 进行中值模糊，去噪点
cimg = cv.cvtColor(img, cv.COLOR_GRAY2BGR)
circles = cv.HoughCircles(img, cv.HOUGH_GRADIENT, 1, 100, param1=100, param2=30, minRadius=30, maxRadius=220)

circles = np.uint16(np.around(circles))
print(circles)

for i in circles[0, :]:  # 遍历矩阵每一行的数据
    cv.circle(planets, (i[0], i[1]), i[2], (0, 255, 0), 3)
    cv.circle(planets, (i[0], i[1]), 2, (0, 255, 0), 3)

cv.imshow("gay_img", planets)
cv.waitKey(0)
cv.destroyAllWindows()